From e0929a675ecc802d6f4a17dc61c17b771471190e Mon Sep 17 00:00:00 2001
From: Pritesh Patel <pritesh.patel@einfochips.com>
Date: Tue, 7 Jan 2025 06:28:33 +0000
Subject: [PATCH 26/74] drivers: cache: sifive_ccache: Enable all available
 ways

Enabled all available cache ways to be used as cache
This change is required for CPU performance improvement.

Signed-off-by: Pritesh Patel <pritesh.patel@einfochips.com>
---
 drivers/cache/sifive_ccache.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/cache/sifive_ccache.c b/drivers/cache/sifive_ccache.c
index 134eac409103..ee6955f64068 100644
--- a/drivers/cache/sifive_ccache.c
+++ b/drivers/cache/sifive_ccache.c
@@ -172,6 +172,16 @@ static const struct riscv_nonstd_cache_ops ccache_mgmt_ops __initconst = {
 };
 #endif /* CONFIG_RISCV_NONSTANDARD_CACHE_OPS */
 
+#ifdef CONFIG_SOC_SIFIVE_EIC7700
+static void ccache_largest_way_enable(void)
+{
+	unsigned int config, ways;
+	config = readl(ccache_base + SIFIVE_CCACHE_CONFIG);
+	ways = (config >> 8) & 0xff;
+	writel(ways-1, ccache_base + SIFIVE_CCACHE_WAYENABLE);
+}
+#endif /* CONFIG_SOC_SIFIVE_EIC7700 */
+
 static int ccache_largest_wayenabled(void)
 {
 	return readl(ccache_base + SIFIVE_CCACHE_WAYENABLE) & 0xFF;
@@ -319,6 +329,11 @@ static int __init sifive_ccache_init(void)
 		goto err_unmap;
 	}
 
+#ifdef CONFIG_SOC_SIFIVE_EIC7700
+	/* Enable all the cache ways */
+	ccache_largest_way_enable();
+#endif
+
 #ifdef CONFIG_RISCV_NONSTANDARD_CACHE_OPS
 	if (quirks & QUIRK_NONSTANDARD_CACHE_OPS) {
 		riscv_cbom_block_size = SIFIVE_CCACHE_LINE_SIZE;
-- 
2.52.0

